pkg/kubelet/images/image_manager.go

```diff
@@ -18,24 +18,49 @@ package images
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"strings"
+	"sync"
 	"time"
 
 	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/client-go/tools/record"
 	"k8s.io/client-go/util/flowcontrol"
 	"k8s.io/klog/v2"
 
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
 	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
 	crierrors "k8s.io/cri-api/pkg/errors"
+	"k8s.io/kubernetes/pkg/credentialprovider"
+	credentialprovidersecrets "k8s.io/kubernetes/pkg/credentialprovider/secrets"
+	"k8s.io/kubernetes/pkg/features"
 	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
 	"k8s.io/kubernetes/pkg/kubelet/events"
 	"k8s.io/kubernetes/pkg/kubelet/metrics"
 	"k8s.io/kubernetes/pkg/util/parsers"
 )
 
+type ensuredInfo struct {
+	// true for ensured secret
+	ensured bool
+	// the secret should be verified again if current time is after the due date.
+	// and the due date is `PullImageSecretRecheckDuration` after last ensured date.
+	// `PullImageSecretRecheckDuration` is configurable in kubelet config.
+	lastEnsuredDate time.Time
+}
+
+type ensureSecretPulledImageDigest struct {
```

Comment:
nit: it's my preference to define structures/functions below where they're used so files can be read top-down. Can this be switched with ensuredInfo?

Reply:
Moved.